### PR TITLE
Arnej/add visit in vespa cli

### PR DESCRIPTION
**Pull request by arnej27959**
this isn't quite finished, but some review etc needed.

- some more description in help text needed
- default is to do nothing with the visited documents, what makes more sense?

@mpolden please review the code
@kkraune any suggestions?

### CODE REVIEW

File: client/go/internal/cli/cmd/visit.go [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+type visitArgs struct {
+	contentCluster string
+	fieldSet       string
+	selection      string
+	makeFeed       bool
+	jsonLines      bool
+	pretty         bool
+	quietMode      bool
+	chunkCount     int
```
Comment (kkraune) [inline client/go/internal/cli/cmd/visit.go:32]: In https://github.com/vespa-engine/sample-apps/pull/1127 is was a little surprise this was not called wantedDocumentCound as in the documentation - confusing

Comment (kkraune) [reply client/go/internal/cli/cmd/visit.go:32]: but chunkCount is a better name, yes :-)

**Review by kkraune — DISMISSED** [review]
The code I tested in my own copy of this, worked well

File: client/go/internal/cli/cmd/visit.go [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+			return nil
+		},
+	}
+	cmd.Flags().StringVar(&vArgs.contentCluster, "content-cluster", "*", `Which content cluster to visit documents from`)
+	cmd.Flags().StringVar(&vArgs.fieldSet, "field-set", "", `Which fieldset to ask for`)
+	cmd.Flags().StringVar(&vArgs.selection, "selection", "", `select subset of cluster`)
+	cmd.Flags().BoolVar(&vArgs.jsonLines, "json-lines", false, `output documents as JSON lines`)
+	cmd.Flags().BoolVar(&vArgs.makeFeed, "make-feed", false, `output JSON array suitable for vespa-feeder`)
+	cmd.Flags().BoolVar(&vArgs.pretty, "pretty-json", false, `format pretty JSON`)
+	cmd.Flags().IntVar(&vArgs.chunkCount, "chunk-count", 1000, `chunk by count`)
```
Comment (kkraune) [inline client/go/internal/cli/cmd/visit.go:80]: this is a great default, thanks! many test with a tiny index and this speeds it up

File: client/go/internal/cli/cmd/visit.go [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+			}
+			result := probeHandler(service)
+			if result.Success {
+				result = visitClusters(vArgs, service)
+			}
+			if !result.Success {
+				return fmt.Errorf("visit failed: %s", result.Message)
+			}
+			if !vArgs.quietMode {
+				fmt.Fprintln(os.Stderr, "[debug] sum of 'documentCount':", totalDocCount)
```
Comment (mpolden) [inline client/go/internal/cli/cmd/visit.go:69]: Consider adding a `printDebug` in the same style as these: https://github.com/vespa-engine/vespa/blob/master/client/go/internal/cli/cmd/root.go#L255-L271

File: client/go/internal/cli/cmd/visit_test.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+package cmd
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestQuoteFunc(t *testing.T) {
```
Comment (mpolden) [inline client/go/internal/cli/cmd/visit_test.go:11]: We should add a test for the visiting part. Something similar is done in https://github.com/vespa-engine/vespa/blob/master/client/go/internal/cli/cmd/document_test.go.

File: client/go/internal/cli/cmd/visit.go [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+	}
+	defer response.Body.Close()
+	vvo, err := parseVisitOutput(response.Body)
+	if response.StatusCode == 200 {
+		if err == nil {
+			totalDocCount += vvo.DocumentCount
+			if vvo.DocumentCount != len(vvo.Documents) {
+				fmt.Fprintln(os.Stderr, "Inconsistent contents from:", url)
+				fmt.Fprintln(os.Stderr, "claimed count: ", vvo.DocumentCount)
+				fmt.Fprintln(os.Stderr, "document blobs: ", len(vvo.Documents))
```
Comment (mpolden) [inline client/go/internal/cli/cmd/visit.go:273]: Consider using https://github.com/vespa-engine/vespa/blob/master/client/go/internal/cli/cmd/root.go#L266

**Review by mpolden — COMMENTED** [review]
Great! But needs more tests.

**Comment by baldersheim** [comment]
@kkraune @mpolden Ready to go ?

**Comment by kkraune** [comment]
I'm OK with this!

**Pull request merged**